polkadot/api/src/lib.rs

```diff
@@ -141,33 +166,44 @@ macro_rules! with_runtime {
 impl<B: Backend> PolkadotApi for Client<B, NativeExecutor<LocalDispatch>>
 	where ::client::error::Error: From<<<B as Backend>::State as state_machine::backend::Backend>::Error>
 {
+	type CheckedBlockId = CheckedId;
 	type BlockBuilder = ClientBlockBuilder<B::State>;
 
-	fn session_keys(&self, at: &BlockId) -> Result<Vec<SessionKey>> {
+	fn check_id(&self, id: BlockId) -> Result<CheckedId> {
+		// bail if the code is not the same as the natively linked.
+		if self.code_at(&id)? != LocalDispatch::native_equivalent() {
+			bail!(ErrorKind::UnknownRuntime);
+		}
+
+		Ok(CheckedId(id))
+	}
+
+	fn session_keys(&self, at: &CheckedId) -> Result<Vec<SessionKey>> {
 		with_runtime!(self, at, ::runtime::consensus::authorities)
 	}
 
-	fn validators(&self, at: &BlockId) -> Result<Vec<AccountId>> {
+	fn validators(&self, at: &CheckedId) -> Result<Vec<AccountId>> {
 		with_runtime!(self, at, ::runtime::session::validators)
 	}
 
-	fn duty_roster(&self, at: &BlockId) -> Result<DutyRoster> {
+	fn duty_roster(&self, at: &CheckedId) -> Result<DutyRoster> {
 		with_runtime!(self, at, ::runtime::parachains::calculate_duty_roster)
 	}
 
-	fn timestamp(&self, at: &BlockId) -> Result<Timestamp> {
+	fn timestamp(&self, at: &CheckedId) -> Result<Timestamp> {
 		with_runtime!(self, at, ::runtime::timestamp::get)
 	}
 
-	fn evaluate_block(&self, at: &BlockId, block: Block) -> Result<()> {
+	fn evaluate_block(&self, at: &CheckedId, block: Block) -> Result<()> {
 		with_runtime!(self, at, || ::runtime::system::internal::execute_block(block))
 	}
 
-	fn build_block(&self, parent: &BlockId, timestamp: Timestamp) -> Result<Self::BlockBuilder> {
-		if self.code_at(parent)? != LocalDispatch::native_equivalent() {
-			bail!(ErrorKind::UnknownRuntime);
-		}
+	fn nonce(&self, at: &Self::CheckedBlockId, account: AccountId) -> Result<TxOrder> {
+		with_runtime!(self, at, || ::runtime::system::nonce(account))
```

Comment:
what ever happened to the plan to allow typed dispatching into wasm? poc-2?

Reply:
filed an issue: https://github.com/paritytech/polkadot/issues/83